faunadb-java-dsl/src/main/java/com/faunadb/client/types/Encoder.java

```diff
@@ -0,0 +1,163 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.types.Properties.Property;
+import com.faunadb.client.types.Value.ArrayV;
+import com.faunadb.client.types.Value.ObjectV;
+
+import java.lang.reflect.Array;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Stack;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Lists.newArrayListWithCapacity;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB object to {@link Value} encoder.
+ */
+public final class Encoder {
+  private Encoder() {
+  }
+
+  private final static ConcurrentHashMap<Class<?>, Codec<?>> CODECS = new ConcurrentHashMap<>();
+
+  static {
+    Codecs.initDefault(CODECS);
+  }
+
+  private final static Stack<Object> stack = new Stack<>();
+
+  /**
+   * Encode the specified object into a corresponding FaunaDB value.
+   * <p>
+   * <pre>{@code
+   * encode(10) => new LongV(10)
+   * encode(3.14) => new DoubleV(3.14)
+   * encode(true) => BooleanV.TRUE
+   * encode(null) => NullV.NULL
+   * encode("a string") => new StringV("a string")
+   * encode(new int[] {1, 2}) => new ArrayV(Arrays.asList(1, 2))
+   * encode(Arrays.asList(1, 2)) => new ArrayV(Arrays.asList(1, 2))
+   * encode(new byte[] {1, 2}) => new ByteV(new byte[] {1, 2})
+   * encode(user) => new ObjectV(ImmutableMap.of("user_name", new StringV("john"), "password", new StringV("s3cr3t")))
+   * }</pre>
+   *
+   * @param obj Any instance of user defined classes, primitive values or any
+   *            generic collection like {@link java.util.List}, {@link java.util.Set} or {@link java.util.Map}
+   * @return A FaunaDB {@link Value} corresponding to the given argument
+   */
+  public static Result<Value> encode(Object obj) {
+    if (obj == null)
+      return Result.<Value>success(Value.NullV.NULL);
+
+    if (Value.class.isInstance(obj))
+      return Result.success((Value) obj);
+
+    if (stack.contains(obj))
```

Comment:
How does `stack` check for equality? If it's using `equals`, two objects might answer true because they have the same content but, they might not be self references.